src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+        let mut hasher = Sha1::new();
+        hasher.input(&data);
+        hasher.result(&mut checksum);
+
+        if checksum == remote_pack.pack_checksum {
+            create_file(pack_path)?.write_all(&data)?;
+        } else {
+            log::error!(
+                "The pack checksum did not match the one in the .esi! The download failed."
+            );
+            return Err(Error::CorruptPack);
+        }
+    }
+    Ok(())
+}
+
+/// Updates all pack index files by fetching the URLs in the [`RemoteIndex`]
+/// only when necessary.
+pub fn update_remote_pack_indexes(
+    remote: &RemoteIndex,
+    base_dir: &Path,
+    reporter: &ProgressReporter,
+) -> Result<(), Error> {
+    let mut done = 0usize;
+    let mut remaining = remote.packs.len();
+
+    for pack in &remote.packs {
+        let url = (pack.url.to_string() + ".idx").parse::<Url>().unwrap();
+        let pack_index_file_name = url.path_segments().unwrap().last().unwrap();
+        let pack_index_path = base_dir.join(pack_index_file_name);
+
+        if verify_checksum(&pack_index_path, &pack.index_checksum) {
+            // The file exists and the checksums match -> skip
+            log::info!("{} is up to date", pack_index_path.display());
+        } else {
+            update_pack_index(&url, &pack_index_path)?;
+        }
+
+        done += 1;
+        remaining -= 1;
+        reporter.checkpoint_with_detail(done, Some(remaining), pack.file_name().to_owned());
+    }
+    Ok(())
+}
+
+/// Updates the pack index file by fetching its contents from the URL only
+/// when the content at the URL is newer than what is available on-disk.
+fn update_pack_index(url: &Url, pack_index_path: &Path) -> Result<(), Error> {
+    let date_modified = fs::metadata(&pack_index_path)
+        .ok()
+        .and_then(|x| x.modified().ok());
+
+    let pack_index_bytes = read_remote_resource(
+        url,
+        Duration::from_secs(15),
+        REMOTE_INDEX_SIZE_LIMIT,
+        date_modified,
+    )?;
+
+    if let Some(pack_index_bytes) = pack_index_bytes {
+        if let Err(e) = PackIndex::parse(pack_index_bytes.as_slice()) {
+            log::error!(
+                "Failed to fetch {} from remote: The remote returned a broken .pack.idx! {}",
+                url,
+                e
+            );
+        } else {
+            log::info!(
+                "Writing {} ({} B)...",
+                pack_index_path.display(),
+                pack_index_bytes.len()
+            );
+            fs::write(&pack_index_path, pack_index_bytes.as_slice()).map_err(Error::IOError)?;
+        }
+    }
+    Ok(())
+}
+
+/// Fetches the remote index from the server.
+pub fn fetch_remote(url: &str, path: &Path) -> Result<RemoteIndex, Error> {
+    let url = url.parse::<Url>().unwrap();
+    let response =
+        read_remote_resource(&url, Duration::from_secs(15), REMOTE_INDEX_SIZE_LIMIT, None)?;
+
+    match response {
+        None => unreachable!(),
```

Comment:
This isn't completely unreachable given a faulty http server or logic bug. At least I think some text should go inside the macro, perhaps something like this? `unreachable!("unexpected not-modified response from server given previously unseen resource")`.

Reply:
Ack.